proposals/multiple_user_script_worlds.md

```diff
@@ -0,0 +1,286 @@
+# Proposal: Allow multiple user script worlds
+
+**Summary**
+
+Allow developers to configure and use multiple user script worlds in the
+`browser.userScripts` API.
+
+**Document Metadata**
+
+**Author:** rdcronin
+
+**Sponsoring Browser:** Google Chrome
+
+**Contributors:** N/A
+
+**Created:** 2024-03-07
+
+**Related Issues:** [565](https://github.com/w3c/webextensions/issues/565)
+
+## Motivation
+
+### Objective
+
+User scripts are (usually small) scripts that enable myriad use cases by
+injecting scripts into different web pages. A user may have many different
+user scripts that execute on the same page; however, these scripts are
+conceptually distinct from one another and should generally not interact with
+each other.
+
+Today, user script managers take various steps to isolate user scripts as they
+are injected, since all user scripts will run either in the context of the
+main world or in a single user script world for the extension. This API would
+allow user script managers to instead have multiple user script worlds, each
+isolated from one another, to reduce the likelihood of user scripts negatively
+affecting one another through collisions.
+
+#### Use Cases
+
+The primary use case is user script managers that handle multiple, independent
+script injections in a single page.
+
+### Known Consumers
+
+We anticipate user script managers such as Tampermonkey, Violentmonkey, and
+others to use this new capability.
+
+## Specification
+
+### Schema
+
+This change would modify existing schemas. We will add a new property,
+`worldId`, to the `WorldProperties` and `RegisteredUserScript` types, as below.
```

Comment:
Q1: I assume that world ids are generated by the browser (when an extension asks to create a new world or world id)?  (as opposed to having an extension generate unique world id strings, and then pass those IDs to the browser/extension APIs)

Q2: Can a malicious extension try to spoof/forge world ids that belong to another extension?

Reply:
Q1: not quite -- these IDs _are_ provided by extensions.  On the chrome side, we'll segment these by extension ID, so that the unique world configuration ID would be pair<extension id, world id> (though note that there may be multiple worlds with that same pair across different renderers -- this is intentional, those should inject in the same world configuration).

Q2: Nope, because of this pairing.  (Extension A's "my world" has no overlap with Extension B's "my world").